File: frontend/src/api/roadmap.ts
import { api } from "./client";

export type RoadmapMonth = {
  month: string;
  label?: string;
  subjects?: Record<
    string,
    {
      focus?: string;
      syllabus_nodes?: string[];
      weekly_hours_hint?: number;
      notes?: string;
    }
  >;
  milestones?: string[];
};

export type StudyRoadmapVersion = {
  id: string;
  version: number;
  source: string;
  start_date: string;
  end_date: string;
  summary_json?: { text?: string } | null;
  months_json: { months?: RoadmapMonth[] };
};

export type StudyRoadmapState = {
  roadmap_id: string | null;
  status: string | null;
  active_version: StudyRoadmapVersion | null;
  pending_version: StudyRoadmapVersion | null;
  generation_job?: {
    id: string;
    status: string;
    error_message?: string | null;
  } | null;
};

export function fetchRoadmap() {
  return api<StudyRoadmapState>("/student/roadmap");
}

export function confirmRoadmap() {
  return api<{ active_version: StudyRoadmapVersion; message: string }>(
    "/student/roadmap/confirm",
    { method: "POST" },
  );
}

export function rejectRoadmap() {
  return api<void>("/student/roadmap/reject", { method: "POST" });
}

File: backend/app/schemas/roadmap.py
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class SyllabusNodeResolvedOut(BaseModel):
    id: str | None = None
    name: str
    parent_name: str | None = None


class RoadmapMonthSubjectOut(BaseModel):
    focus: str = ""
    syllabus_nodes: list[str] = Field(default_factory=list)
    syllabus_node_ids: list[str] = Field(default_factory=list)
    syllabus_nodes_resolved: list[SyllabusNodeResolvedOut] = Field(default_factory=list)
    weekly_hours_hint: int | None = None
    notes: str = ""


class RoadmapMonthOut(BaseModel):
    month: str
    label: str = ""
    subjects: dict[str, RoadmapMonthSubjectOut] = Field(default_factory=dict)
    milestones: list[str] = Field(default_factory=list)


class StudyRoadmapVersionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    version: int
    source: str
    start_date: date
    end_date: date
    summary_json: dict | None = None
    months_json: dict
    created_at: datetime


class RoadmapGenerationJobOut(BaseModel):
    id: str
    status: str
    error_message: str | None = None
    roadmap_version_id: str | None = None


class StudyRoadmapStateOut(BaseModel):
    roadmap_id: uuid.UUID | None = None
    status: str | None = None
    active_version: StudyRoadmapVersionOut | None = None
    pending_version: StudyRoadmapVersionOut | None = None
    generation_job: RoadmapGenerationJobOut | None = None


class RoadmapConfirmOut(BaseModel):
    active_version: StudyRoadmapVersionOut
    message: str = "全年计划已生效"


class RoadmapRegenerateOut(BaseModel):
    job_id: uuid.UUID
    created: bool
